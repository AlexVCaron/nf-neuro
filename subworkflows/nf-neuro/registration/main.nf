include { REGISTRATION_ANATTODWI  } from '../../../modules/nf-neuro/registration/anattodwi/main'
include { REGISTRATION_ANTS   } from '../../../modules/nf-neuro/registration/ants/main'
include { REGISTRATION_EASYREG   } from '../../../modules/nf-neuro/registration/easyreg/main'
include { REGISTRATION_SYNTHREGISTRATION } from '../../../modules/nf-neuro/registration/synthregistration/main'
include { REGISTRATION_CONVERT } from '../../../modules/nf-neuro/registration/convert/main'

params.run_easyreg      = false
params.run_synthmorph   = false


workflow REGISTRATION {

    // The subworkflow requires at least ch_fixed_image and ch_moving_image as inputs to
    // properly perform the registration. Supplying a ch_metric will select
    // the REGISTRATION_ANATTODWI module meanwhile NOT supplying a ch_metric
    // will select the REGISTRATION_ANTS (SyN or SyNQuick) module. Alternatively,
    // NOT supplying ch_metric and activating alternative module flag with select
    // REGISTRATION_EASYREG or REGISTRATION_SYNTHMORPH

    take:
        ch_fixed_image                  // channel: [ val(meta), image ]
        ch_moving_image                 // channel: [ val(meta), reference ]
        ch_metric                       // channel: [ val(meta), metric ], optional
        ch_fixed_mask                   // channel: [ val(meta), mask ], optional
        ch_segmentation                 // channel: [ val(meta), segmentation ], optional
        ch_moving_segmentation          // channel: [ val(meta), segmentation ], optional
        ch_freesurfer_license           // channel: [ license ], optional
    main:
        ch_versions = Channel.empty()
        ch_mqc = Channel.empty()

        if ( params.run_easyreg ) {
            // ** Registration using Easyreg ** //
            // Result : [ meta, reference, image | [], ref-segmentation | [], segmentation | [] ]
            //  Steps :
            //   - join [ meta, reference, image | null ]
            //   - join [ meta, reference, image | null, ref-segmentation | null ]
            //   - join [ meta, reference, image | null, ref-segmentation | null, segmentation | null ]
            //   -  map [ meta, reference, image | [], ref-segmentation | [], segmentation | [] ]
            ch_register = ch_moving_image
                .join(ch_fixed_image, remainder: true)
                .join(ch_moving_segmentation, remainder: true)
                .join(ch_segmentation, remainder: true)
                .map{ it[0..1] + [it[2] ?: [], it[3] ?: [], it[4] ?: []] }

            REGISTRATION_EASYREG ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_EASYREG.out.versions.first())

            // ** Set compulsory outputs ** //
            out_image_warped = REGISTRATION_EASYREG.out.image_warped
            out_affine = Channel.empty()
            out_warp = REGISTRATION_EASYREG.out.warp
            out_inverse_affine = Channel.empty()
            out_inverse_warp = REGISTRATION_EASYREG.out.inverse_warp
            out_image_transform = REGISTRATION_EASYREG.out.warp
            out_inverse_image_transform = REGISTRATION_EASYREG.out.inverse_warp
            out_tractogram_transform = REGISTRATION_EASYREG.out.inverse_warp
            out_inverse_tractogram_transform = REGISTRATION_EASYREG.out.warp

            // ** Set optional outputs. ** //
            // If segmentations are not provided as inputs,
            // easyreg will outputs synthseg segmentations
            out_ref_warped = REGISTRATION_EASYREG.out.fixed_warped
            out_segmentation = ch_segmentation.mix( REGISTRATION_EASYREG.out.segmentation_warped )
            out_ref_segmentation = ch_moving_segmentation.mix( REGISTRATION_EASYREG.out.fixed_segmentation_warped )
        }
        else if ( params.run_synthmorph ) {
            // ** Registration using synthmorph ** //
            ch_register = ch_fixed_image
                .join(ch_moving_image)

            REGISTRATION_SYNTHREGISTRATION ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_SYNTHREGISTRATION.out.versions.first())

            // Tag all synthmorph transforms per type, and index if in a chain. This info will be
            // used after conversion to sort out the transforms from the conversion module.
            ch_convert_affine = REGISTRATION_SYNTHREGISTRATION.out.affine
                .map{ meta, affine -> [meta, [tag: "affine"], affine] }
            ch_convert_warp = REGISTRATION_SYNTHREGISTRATION.out.warp
                .map{ meta, warp -> [meta, [tag: "warp"], warp] }
            ch_convert_inverse_affine = REGISTRATION_SYNTHREGISTRATION.out.inverse_affine
                .map{ meta, inverse_affine -> [meta, [tag: "inverse_affine"], inverse_affine] }
            ch_convert_inverse_warp = REGISTRATION_SYNTHREGISTRATION.out.inverse_warp
                .map{ meta, inverse_warp -> [meta, [tag: "inverse_warp"], inverse_warp] }
            ch_convert_image_transform = REGISTRATION_SYNTHREGISTRATION.out.image_transform
                .map{ meta, transforms -> [meta, [tag: "image_transform"], 0..<transforms.size(), transforms] }
                .transpose()
                .map{ meta, tag, idx, transform -> [meta, tag + [idx: idx], transform]}
            ch_convert_inverse_image_transform = REGISTRATION_SYNTHREGISTRATION.out.inverse_image_transform
                .map{ meta, transforms -> [meta, [tag: "inverse_image_transform"], 0..<transforms.size(), transforms] }
                .transpose()
                .map{ meta, tag, idx, transform -> [meta, tag + [idx: idx], transform]}

            // Mix all transforms into a single channel for conversion
            ch_convert = ch_convert_affine.view{ "affine $it" }
                .mix(ch_convert_warp.view{ "warp $it" })
                .mix(ch_convert_inverse_affine.view{ "inv-affine $it" })
                .mix(ch_convert_inverse_warp.view{ "inv-warp $it" })
                .mix(ch_convert_image_transform.view{ "image-transform $it" })
                .mix(ch_convert_inverse_image_transform.view{ "inv-image-transform $it" })
                .combine(ch_fixed_image, by: 0)
                .combine(ch_moving_image, by: 0)
                .map{ meta, tag, transform, fixed, moving ->
                    def extension = transform.name.tokenize('.')[1..-1].join(".")
                    return [
                        meta + tag + [cache: meta],
                        transform,
                        extension == "lta" ? "lta" : "ras",
                        "itk",
                        extension == "lta" ? fixed : moving,
                        [],
                    ]}
                .combine(ch_freesurfer_license)

            REGISTRATION_CONVERT ( ch_convert )
            ch_versions = ch_versions.mix(REGISTRATION_CONVERT.out.versions.first())

            // Un-mix conversion outputs using the tags. Save indexes for output sorting
            ch_conversion_outputs = REGISTRATION_CONVERT.out.transformation
                .branch{ meta, transform ->
                    affine: meta.tag == "affine"
                        return [meta.cache, transform]
                    warp: meta.tag == "warp"
                        return [meta.cache, transform]
                    inverse_affine: meta.tag == "inverse_affine"
                        return [meta.cache, transform]
                    inverse_warp: meta.tag == "inverse_warp"
                        return [meta.cache, transform]
                    image_transform: meta.tag == "image_transform"
                        return [meta.cache, [idx: meta.idx, trans: transform]]
                    inverse_image_transform: meta.tag == "inverse_image_transform"
                        return [meta.cache, [idx: meta.idx, trans: transform]]
                }


            // ** Set compulsory outputs ** //
            out_image_warped = REGISTRATION_SYNTHREGISTRATION.out.image_warped
            out_affine = ch_conversion_outputs.affine
            out_warp = ch_conversion_outputs.warp
            out_inverse_affine = ch_conversion_outputs.inverse_affine
            out_inverse_warp = ch_conversion_outputs.inverse_warp
            out_image_transform = ch_conversion_outputs.image_transform
                .groupTuple()
                .map{ meta, trans -> [meta, trans.sort{ t1, t2 -> t1.idx <=> t2.idx }.collect{ it.trans }] }
            out_inverse_image_transform = ch_conversion_outputs.inverse_image_transform
                .groupTuple()
                .map{ meta, trans -> [meta, trans.sort{ t1, t2 -> t1.idx <=> t2.idx }.collect{ it.trans }] }
            out_tractogram_transform = out_inverse_image_transform
            out_inverse_tractogram_transform = out_image_transform
            // ** and optional outputs. ** //
            out_ref_warped = Channel.empty()
            out_segmentation = Channel.empty()
            out_ref_segmentation = Channel.empty()
        }
        else {
            // ** Classic registration using antsRegistration  ** //
            // Result : [ meta, image, reference, metric | [] ]
            //  Steps :
            //   - join [ meta, image, ref ]
            //   - join [ meta, image, ref, metric | null ]
            //   - map  [ meta, image, ref, metric | [] ]
            // Branches :
            //   - anat_to_dwi : has a metric at index 3
            //   - ants_syn    : doesn't have a metric at index 3 ( [] or null )
            ch_register = ch_fixed_image
                .join(ch_moving_image)
                .join(ch_metric, remainder: true)
                .map{ it[0..2] + [it[3] ?: []] }
                .branch{
                    anat_to_dwi : it[3]
                    ants_syn: true
                }

            // ** Registration using ANAT TO DWI ** //
            REGISTRATION_ANATTODWI ( ch_register.anat_to_dwi )
            ch_versions = ch_versions.mix(REGISTRATION_ANATTODWI.out.versions.first())
            ch_mqc = ch_mqc.mix(REGISTRATION_ANATTODWI.out.mqc)

            // ** Set compulsory outputs ** //
            out_image_warped = REGISTRATION_ANATTODWI.out.anat_warped
            out_affine = REGISTRATION_ANATTODWI.out.affine
            out_warp = REGISTRATION_ANATTODWI.out.warp
            out_inverse_affine = REGISTRATION_ANATTODWI.out.inverse_affine
            out_inverse_warp = REGISTRATION_ANATTODWI.out.inverse_warp
            out_image_transform = REGISTRATION_ANATTODWI.out.image_transform
            out_inverse_image_transform = REGISTRATION_ANATTODWI.out.inverse_image_transform
            out_tractogram_transform = REGISTRATION_ANATTODWI.out.tractogram_transform
            out_inverse_tractogram_transform = REGISTRATION_ANATTODWI.out.inverse_tractogram_transform

            // ** Registration using ANTS SYN SCRIPTS ** //
            // Registration using antsRegistrationSyN.sh or antsRegistrationSyNQuick.sh, has
            // to be defined in the config file or else the default (SyN) will be used.
            // Result : [ meta, image, mask | [] ]
            //  Steps :
            //   - join [ meta, image, metric | [], mask | null ]
            //   - map  [ meta, image, mask | [] ]
            ch_register = ch_register.ants_syn
                .join(ch_fixed_mask, remainder: true)
                .map{ it[0..2] + [it[4] ?: []] }

            REGISTRATION_ANTS ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_ANTS.out.versions.first())
            ch_mqc = ch_mqc.mix(REGISTRATION_ANTS.out.mqc)

            // ** Set compulsory outputs ** //
            out_image_warped = out_image_warped.mix(REGISTRATION_ANTS.out.image_warped)
            out_affine = out_affine.mix(REGISTRATION_ANTS.out.affine)
            out_warp = out_warp.mix(REGISTRATION_ANTS.out.warp)
            out_inverse_affine = out_inverse_affine.mix(REGISTRATION_ANTS.out.inverse_affine)
            out_inverse_warp = out_inverse_warp.mix(REGISTRATION_ANTS.out.inverse_warp)
            out_image_transform = out_image_transform.mix(REGISTRATION_ANTS.out.image_transform)
            out_inverse_image_transform = out_inverse_image_transform.mix(REGISTRATION_ANTS.out.inverse_image_transform)
            out_tractogram_transform = out_tractogram_transform.mix(REGISTRATION_ANTS.out.tractogram_transform)
            out_inverse_tractogram_transform = out_inverse_tractogram_transform.mix(REGISTRATION_ANTS.out.inverse_tractogram_transform)

            // **and optional outputs **//
            out_ref_warped = Channel.empty()
            out_segmentation = Channel.empty()
            out_ref_segmentation = Channel.empty()
        }
    emit:
        image_warped                    = out_image_warped                  // channel: [ val(meta), image ]
        ref_warped                      = out_ref_warped                    // channel: [ val(meta), ref ]
        // Individual transforms
        affine                          = out_affine                        // channel: [ val(meta), <affine> ]
        warp                            = out_warp                          // channel: [ val(meta), <warp> ]
        inverse_warp                    = out_inverse_warp                  // channel: [ val(meta), <inverse-warp> ]
        inverse_affine                  = out_inverse_affine                // channel: [ val(meta), <inverse-affine> ]
        // Combined transforms
        image_transform                 = out_image_transform               // channel: [ val(meta), [ <warp>, <affine> ] ]
        inverse_image_transform         = out_inverse_image_transform       // channel: [ val(meta), [ <inverse-affine>, <inverse-warp> ] ]
        tractogram_transform            = out_tractogram_transform          // channel: [ val(meta), [ <inverse-affine>, <inverse-warp> ] ]
        inverse_tractogram_transform    = out_inverse_tractogram_transform  // channel: [ val(meta), [ <warp>, <affine> ] ]
        // Segmentations
        segmentation                    = out_segmentation                  // channel: [ val(meta), segmentation ]
        ref_segmentation                = out_ref_segmentation              // channel: [ val(meta), ref-segmentation ]

        mqc                             = ch_mqc                            // channel: [ *mqc.*, ... ]
        versions                        = ch_versions                       // channel: [ versions.yml ]
}
