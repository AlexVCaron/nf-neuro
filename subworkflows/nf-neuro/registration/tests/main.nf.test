nextflow_workflow {

    name "Test Subworkflow REGISTRATION"
    script "../main.nf"
    workflow "REGISTRATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/registration"

    tag "registration"
    tag "registration/anattodwi"
    tag "registration"
    tag "registration/ants"
    tag "registration/easyreg"
    tag "registration/synthregistration"

    tag "load_test_data"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "T1w.zip", "b0.zip", "dti.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("registration - antsRegistration") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[2] = ch_split_test_data.dti.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz")
                    ]
                }
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() }
            )
        }
    }

    test("registration - SyNQuick") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]}
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() }
            )
        }
    }

    test("registration - easyreg") {
        config "./nextflow_easyreg.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() }
            )
        }
    }
    test("registration - synthregistration") {
        config "./nextflow_synthregistration.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        b0: it.simpleName == "b0"
                        dti: it.simpleName == "dti"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ !it.key.isInteger() && it.value  }
                        .collect{ item -> ["versions"].contains(item.key)
                            ? item.value
                            : file(item.value.get(0).get(1)).name}
                ).match() }
            )
        }
    }
}
