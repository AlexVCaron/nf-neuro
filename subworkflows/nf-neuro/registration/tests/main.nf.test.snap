{
    "registration - Synthmorph": {
        "content": [
            {
                "backward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.txt"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.txt",
                        "test_out_warp.nii.gz"
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_warp.nii.gz",
                        "test_out_affine.txt"
                    ]
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.txt"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_warp.nii.gz",
                        "test_out_affine.txt"
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_out_affine.txt",
                        "test_out_warp.nii.gz"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,300199ef2ea6dd6438553c3ec2615e58",
                    "versions.yml:md5,5bc0d347809508f4b94ebe433b0f21b9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T20:51:47.326447932"
    },
    "registration - easyreg": {
        "content": [
            {
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_warp.nii.gz"
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz"
                    ]
                ],
                "backward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_warp.nii.gz"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz"
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_warp.nii.gz"
                    ]
                ],
                "forward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped.nii.gz"
                    ]
                ],
                "reference_segmentation": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped_reference_segmentation.nii.gz"
                    ]
                ],
                "reference_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped_reference.nii.gz"
                    ]
                ],
                "segmentation": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped_segmentation.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3f38c911476c605480c02e484fcf6e9f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-15T18:55:55.830874285"
    },
    "registration - ANTs - SyNQuick": {
        "content": [
            {
                "backward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat",
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz",
                        "test_forward1_affine.mat"
                    ]
                ],
                "backward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz",
                        "test_forward1_affine.mat"
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat",
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "forward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_ants_mqc.gif"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5f3f8c831b55885ffcc7bd976939dbad"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-16T15:27:00.460253907"
    },
    "registration - ANTs - Anat to DWI": {
        "content": [
            {
                "affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__forward1_affine.mat"
                    ]
                ],
                "image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__forward0_warp.nii.gz",
                        "test__forward1_affine.mat"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test__T1w_warped.nii.gz"
                    ]
                ],
                "inverse_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__backward0_affine.mat"
                    ]
                ],
                "inverse_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__backward0_affine.mat",
                        "test__backward1_warp.nii.gz"
                    ]
                ],
                "inverse_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__forward0_warp.nii.gz",
                        "test__forward1_affine.mat"
                    ]
                ],
                "inverse_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test__backward1_warp.nii.gz"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_anattodwi_mqc.gif"
                    ]
                ],
                "tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__backward0_affine.mat",
                        "test__backward1_warp.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0f0af6043cb0549dd850d1f59d0c7190"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-17T17:41:51.560683985"
    },
    "registration - SyNQuick": {
        "content": [
            {
                "affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__forward1_affine.mat"
                    ]
                ],
                "image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test__forward0_warp.nii.gz",
                        "test__forward1_affine.mat"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_warped.nii.gz"
                    ]
                ],
                "inverse_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat"
                    ]
                ],
                "backward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat",
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "backward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz",
                        "test_forward1_affine.mat"
                    ]
                ],
                "backward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "forward_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward1_affine.mat"
                    ]
                ],
                "forward_image_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz",
                        "test_forward1_affine.mat"
                    ]
                ],
                "forward_tractogram_transform": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat",
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "forward_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz"
                    ]
                ],
                "image_warped": [
                    [
                        {
                            "id": "test"
                        },
                        "test_T1w_warped.nii.gz"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_registration_anattodwi_mqc.gif"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,42f2dcb95a2cf29b157dbaebf162d2dd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-17T17:41:57.285829729"
    }
}